### PR TITLE
Temporary object pools. Fixes and tuning

### PR DESCRIPTION
**Pull request by k-danil**
Good day!

This commit introduce sync.pool's for temporary object and few other optimisations:
- Pool for packet slice between packet_pool.add & data.parseData.
- Pool for payload slice in data.parseData. 
- Copy FirstPacket to DemuxerData without payload. I'm planning to add optional pool for packets and this change is necessary for it.
- Make DemuxerData slice of Sections size.
- Use map[uint32] instead map[uint16] as go runtime provide optimised hash functions for (u)int32/64 keys. This change almost double up map throughput by excluding AES hashing. 
- Remove sync.mutex from programMap and packetPool. This code can not be asynchronous, so mutex usage is redundant.
- Fix BenchmarkParsePSIData & BenchmarkDemuxer_NextData.

Profiling data from our app (we aren't interested in PSI parsing, so it's omitted)

Current master:
before
This PR:
after

Also GC pressure drops on 1/4.

PS PacketSlice pool removed from PR as pooling the slice of references may lead to memory leaks.

Sorry for PR-attack ;) There will be one or two more to implement optional pooling for packets and pes in demuxer.

### CODE REVIEW

**Comment by k-danil** [comment]
Good day, @asticode ! Do you have time to review this PR?

**Comment by asticode** [comment]
Hey, I'll try to take a look next week 👍

File: object_pool.go [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+}
+
+// poolTempPayload is a pool for temporary payload in parseData()
+// Don't use it anywhere else to avoid pool pollution
+type poolTempPayload struct {
+	sp sync.Pool
+}
+
+// get returns the tempPayload object with byte slice of a 'size' length
+func (ptp *poolTempPayload) get(size int) (payload *tempPayload) {
```
Comment (asticode) [inline object_pool.go:29]: Which difference does this function make? Isn't the slice automatically grown when needed?

File: program_map.go [hunk 2]
```diff
@@ -1,54 +1,41 @@
 package astits
 
-import "sync"
-
 // programMap represents a program ids map
 type programMap struct {
-	m *sync.RWMutex
-	p map[uint16]uint16 // map[ProgramMapID]ProgramNumber
+	p map[uint32]uint16 // map[ProgramMapID]ProgramNumber
```
Comment (asticode) [inline program_map.go:5]: What is the metrics difference made by switching from uint32 to uint16 here?

File: object_pool.go (hunk 1 above)
Comment (k-danil) [reply object_pool.go:29]: 1) We are using this slice to copy existent payload prom packets, so we must be sure that the destination slice length is appropriate 
2) We can replace `copy` with `append`, but if capacity of destination slice is lesser than target length it will lead to grow (potentially multiple grow if pooled slice is to small (e.g. reusing slice for PES after initially using it for PSI)). Also this change will not lift our obligations to reset pooled slice to `[:0]`
3) In this function we growing or trimming the slice to appropriate target length in one operation. It can be seen as drop off replacement for `payload = make([]byte, l)` where we creating slice of known length

File: program_map.go [hunk 3]
```diff
@@ -1,54 +1,41 @@
 package astits
 
-import "sync"
-
 // programMap represents a program ids map
 type programMap struct {
-	m *sync.RWMutex
```
[the remark this answers is not in the document]
Comment (k-danil) [reply program_map.go:7]: Done

File: program_map.go (hunk 2 above)
Comment (k-danil) [reply program_map.go:5]: [Benchmark](https://gist.github.com/k-danil/37b52bc9dd9f2e8c9c92119a6aa6c514) gave me following results:
```
BenchmarkMaps
BenchmarkMaps/uint8
BenchmarkMaps/uint8-8         	 2317094	       505.0 ns/op	       0 B/op	       0 allocs/op
BenchmarkMaps/uint16
BenchmarkMaps/uint16-8        	 2190326	       548.1 ns/op	       0 B/op	       0 allocs/op
BenchmarkMaps/uint32
BenchmarkMaps/uint32-8        	 3035148	       395.3 ns/op	       0 B/op	       0 allocs/op
BenchmarkMaps/uint64
BenchmarkMaps/uint64-8        	 3026098	       396.4 ns/op	       0 B/op	       0 allocs/op
PASS
```
This will be absolutely neglectable, but every packet should be passed to packetPool and then checked via programMap in packetAccumulator.add().
Profiling data is kind of unstable (600sec run on 10mbps stream): 
- packetPool: before 90-120ms after 30-50ms
- programMap: befor 20-30ms after unmeasurable

File: muxer.go [hunk 4]
```diff
@@ -45,7 +45,7 @@ type Muxer struct {
 	buf       bytes.Buffer
 	bufWriter *astikit.BitsWriter
 
-	esContexts              map[uint16]*esContext
+	esContexts              map[uint32]*esContext
```
Comment (asticode) [inline muxer.go:48]: Could you add the following comment on top of `esContexts`: `// We use map[uint32] instead map[uint16] as go runtime provide optimized hash functions for (u)int32/64 keys`

File: packet_pool.go [hunk 5]
```diff
@@ -56,24 +60,22 @@ func (b *packetAccumulator) add(p *Packet) (ps []*Packet) {
 
 // packetPool represents a queue of packets for each PID in the stream
 type packetPool struct {
-	b map[uint16]*packetAccumulator // Indexed by PID
-	m *sync.Mutex
+	b map[uint32]*packetAccumulator // Indexed by PID
```
Comment (asticode) [inline packet_pool.go:63]: Can you add the following comment on top of `b`: `// We use map[uint32] instead map[uint16] as go runtime provide optimized hash functions for (u)int32/64 keys`

File: program_map.go (hunk 2 above)
Comment (asticode) [inline program_map.go:5]: Can you add the following comment on top of `p` : `// We use map[uint32] instead map[uint16] as go runtime provide optimized hash functions for (u)int32/64 keys`

File: object_pool.go [hunk 6]
```diff
@@ -0,0 +1,45 @@
+package astits
```
Comment (asticode) [inline object_pool.go:1]: Could you rename the file to `bytes_pool.go`?

File: object_pool.go [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+
+import "sync"
+
+// poolOfTempPayload global variable is used to ease access to pool from any place of the code
+var poolOfTempPayload = &poolTempPayload{
```
Comment (asticode) [inline object_pool.go:6]: could you rename the var to `bytesPool`?

File: object_pool.go [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+}
+
+// tempPayload is an object containing payload slice
+type tempPayload struct {
+	s []byte
+}
+
+// poolTempPayload is a pool for temporary payload in parseData()
+// Don't use it anywhere else to avoid pool pollution
+type poolTempPayload struct {
```
Comment (asticode) [inline object_pool.go:24]: could you rename this struct to `bytesPooler`?

File: object_pool.go [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+			// Prepare the slice of somewhat sensible initial size to minimize calls to runtime.growslice
+			return &tempPayload{
+				s: make([]byte, 0, 1024),
+			}
+		},
+	},
+}
+
+// tempPayload is an object containing payload slice
+type tempPayload struct {
```
Comment (asticode) [inline object_pool.go:18]: Why not use []byte directly instead of a struct containing []byte?

Comment (k-danil) [reply object_pool.go:18]: Sorry for Japanese [resource](https://qiita.com/peroxyacyl/items/5e02ddf4480ecd2ec7b1) but code and benchmarks are still in English.
tl;dr We can pass slices directly, but it lead to additional allocations, since sync.pools are optimised to work with pointers. We can pass *[]byte object from/to pool, but we should track original pointer and update it accordingly. 

In my opinion, it's much simpler and safer to wrap []byte in some struct or use &bytes.Buffer{}.

File: object_pool.go (hunk 6 above)
Comment (k-danil) [reply object_pool.go:1]: I planned to use this file for future implementation of *Packet and *PESData pools. If you think that it should be separated, i will do it.

File: muxer.go (hunk 4 above)
Comment (k-danil) [reply muxer.go:49]: Done

File: packet_pool.go (hunk 5 above)
Comment (k-danil) [reply packet_pool.go:64]: Done

File: program_map.go (hunk 2 above)
Comment (k-danil) [reply program_map.go:6]: Done

File: object_pool.go (hunk 7 above)
Comment (k-danil) [reply object_pool.go]: Done

File: object_pool.go (hunk 8 above)
Comment (k-danil) [reply object_pool.go]: Done

File: object_pool.go (hunk 9 above)
Comment (k-danil) [reply object_pool.go:18]: Additionally there is good Chinese [article](https://blog.csdn.net/FishGoddess/article/details/118254178) ([translation](https://programs.wiki/wiki/an-exploration-caused-by-a-pit-in-go-object-pool.html)) on this topic.

Comment (asticode) [reply object_pool.go:18]: OK in that case could you rename this struct to `bytesPoolItem`?

File: object_pool.go (hunk 6 above)
Comment (asticode) [reply object_pool.go:1]: Mmm keeping them in the same file makes sense, could you just rename the file to `pools.go`?

**Review by asticode — CHANGES_REQUESTED** [review]
Last 2 changes needed, we're all good after that 👍

File: object_pool.go (hunk 6 above)
Comment (k-danil) [reply object_pool.go]: Done

File: object_pool.go (hunk 9 above)
Comment (k-danil) [reply object_pool.go]: Done

**Comment by asticode** [comment]
Thanks for the PR 👍
